core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -309,6 +321,104 @@ private void balance() {
+                            || (subDecision.type() == Type.THROTTLE && explanation == null)) {
+                        explanation = subDecision.getExplanation();
+                        explanationCause = subDecision.type();
+                    }
+                }
+                return new RebalanceDecision(canRebalance, Type.NO, explanation);
+            }
+
+            if (allocation.hasPendingAsyncFetch()) {
+                return new RebalanceDecision(
+                    canRebalance,
+                    Type.NO,
+                    "cannot rebalance due to in-flight shard store fetches, otherwise allocation may prematurely rebalance a shard to " +
+                        "a node that is soon to receive another shard assignment upon completion of the shard store fetch, " +
+                        "rendering the cluster imbalanced again"
+                );
+            }
+
+            sorter.reset(shard.getIndexName());
+            ModelNode[] modelNodes = sorter.modelNodes;
+            final String currentNodeId = shard.currentNodeId();
+            // find currently assigned node
+            ModelNode currentNode = null;
+            for (ModelNode node : modelNodes) {
+                if (node.getNodeId().equals(currentNodeId)) {
+                    currentNode = node;
+                    break;
+                }
+            }
+            assert currentNode != null : "currently assigned node could not be found";
+
+            // balance the shard, if a better node can be found
+            final float currentWeight = sorter.weight(currentNode);
+            final AllocationDeciders deciders = allocation.deciders();
+            final String idxName = shard.getIndexName();
+            Map<String, NodeRebalanceDecision> nodeDecisions = new HashMap<>();
+            Type rebalanceDecisionType = Type.NO;
+            String assignedNodeId = null;
+            for (ModelNode node : modelNodes) {
+                if (node == currentNode) {
+                    continue; // skip over node we're currently allocated to it
+                }
+                final Decision canAllocate = deciders.canAllocate(shard, node.getRoutingNode(), allocation);
+                final float nodeWeight = sorter.weight(node);
+                final boolean betterWeightThanCurrent = nodeWeight <= currentWeight;
+                final boolean rebalanceConditionsMet;
+                final boolean deltaAboveThreshold;
+                final float weightWithShardAdded;
+                if (betterWeightThanCurrent) {
+                    final float currentDelta = absDelta(nodeWeight, currentWeight);
+                    deltaAboveThreshold = lessThan(currentDelta, threshold) == false;
+                    weightWithShardAdded = weight.weightShardAdded(this, node, idxName);
+                    final float proposedDelta = weightWithShardAdded - weight.weightShardRemoved(this, currentNode, idxName);
```

Comment:
I feel like these methods (the comparisons, deltas, etc), would benefit from some comments explaining what's happening for people reading this code